File: components/gov-chat-frontend/src/__tests__/components/ds/DsStateDisplay.test.js
import { mount } from '@vue/test-utils';
import DsStateDisplay from '@/components/ds/StateDisplay.vue';

const mountState = (props = {}, slots = {}) => {
  return mount(DsStateDisplay, { props, slots });
};

describe('DsStateDisplay (AC12)', () => {
  describe('Empty state', () => {
    it('renders empty type by default', () => {
      const wrapper = mountState();
      expect(wrapper.classes()).toContain('ds-state--empty');
    });

    it('renders message text', () => {
      const wrapper = mountState({ message: 'Nothing here' });
      expect(wrapper.find('.ds-state__message').text()).toBe('Nothing here');
    });
  });

  describe('Loading state', () => {
    it('renders loading type', () => {
      const wrapper = mountState({ type: 'loading' });
      expect(wrapper.classes()).toContain('ds-state--loading');
    });
  });

  describe('Error state', () => {
    it('renders error type', () => {
      const wrapper = mountState({ type: 'error' });
      expect(wrapper.classes()).toContain('ds-state--error');
    });
  });

  describe('Message rendering', () => {
    it('does not render message element when no message or slot', () => {
      const wrapper = mountState();
      expect(wrapper.find('.ds-state__message').exists()).toBe(false);
    });

    it('renders message from prop', () => {
      const wrapper = mountState({ message: 'An error occurred' });
      expect(wrapper.find('.ds-state__message').text()).toBe('An error occurred');
    });

    it('renders default slot as message', () => {
      const wrapper = mountState({}, { default: 'Custom message' });
      expect(wrapper.find('.ds-state__message').text()).toBe('Custom message');
    });
  });

  describe('Icon slot', () => {
    it('renders icon slot when provided', () => {
      const wrapper = mountState({}, { icon: '<svg class="icon"></svg>' });
      expect(wrapper.find('.ds-state__icon').exists()).toBe(true);
      expect(wrapper.find('.icon').exists()).toBe(true);
    });

    it('does not render icon section when no icon slot or prop', () => {
      const wrapper = mountState();
      expect(wrapper.find('.ds-state__icon').exists()).toBe(false);
    });
  });

  describe('Action slot', () => {
    it('renders action slot when provided', () => {
      const wrapper = mountState({}, { action: '<button>Retry</button>' });
      expect(wrapper.find('.ds-state__action').exists()).toBe(true);
      expect(wrapper.find('.ds-state__action').text()).toBe('Retry');
    });

    it('does not render action section when no action slot', () => {
      const wrapper = mountState();
      expect(wrapper.find('.ds-state__action').exists()).toBe(false);
    });
  });

  it('has base ds-state class', () => {
    const wrapper = mountState();
    expect(wrapper.classes()).toContain('ds-state');
  });
});

File: mobile/genie_ai_mobile/test/design_system/components/ds_state_display_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:genie_ai_mobile/design_system/components/ds_button.dart';
import 'package:genie_ai_mobile/design_system/components/ds_state_display.dart';

import '../../helpers/test_app.dart';
import '../../helpers/theme_helper.dart';

void main() {
  setUp(setupLightTokens);
  tearDown(resetThemeManager);

  group('DsStateDisplay', () {
    group('loading type', () {
      testWidgets('renders DsSpinner by default', (tester) async {
        await tester.pumpWidget(
          testApp(DsStateDisplay(type: DsStateType.loading)),
        );
        expect(find.byKey(const ValueKey('ds-state-spinner')), findsOneWidget);
      });

      testWidgets('renders customChild when provided', (tester) async {
        await tester.pumpWidget(
          testApp(
            DsStateDisplay(
              type: DsStateType.loading,
              customChild: Text('Loading...', key: const ValueKey('custom-loader')),
            ),
          ),
        );
        final child = tester.widget<Text>(
          find.byKey(const ValueKey('custom-loader')),
        );
        expect(child.data, 'Loading...');
        expect(find.byKey(const ValueKey('ds-state-spinner')), findsNothing);
      });
    });

    group('empty type', () {
      testWidgets('renders default icon and message', (tester) async {
        await tester.pumpWidget(
          testApp(DsStateDisplay(type: DsStateType.empty)),
        );
        final icon = tester.widget<Icon>(
          find.byKey(const ValueKey('ds-state-icon')),
        );
        expect(icon.icon, Icons.inbox_outlined);
        final message = tester.widget<Text>(
          find.byKey(const ValueKey('ds-state-message')),
        );
        expect(message.data, 'No data');
      });

      testWidgets('renders custom message', (tester) async {
        await tester.pumpWidget(
          testApp(
            DsStateDisplay(
              type: DsStateType.empty,
              message: 'Nothing here',
            ),
          ),
        );
        final message = tester.widget<Text>(
          find.byKey(const ValueKey('ds-state-message')),
        );
        expect(message.data, 'Nothing here');
      });

      testWidgets('renders action button when provided', (tester) async {
        var actionFired = false;
        await tester.pumpWidget(
          testApp(
            DsStateDisplay(
              type: DsStateType.empty,
              actionLabel: 'Retry',
              onAction: () => actionFired = true,
            ),
          ),
        );
        final actionLabel = tester.widget<Text>(
          find.byKey(const ValueKey('ds-button-label')),
        );
        expect(actionLabel.data, 'Retry');
        await tester.tap(find.byKey(const ValueKey('ds-state-action')));
        expect(actionFired, isTrue);
      });
    });

    group('error type', () {
      testWidgets('renders danger icon and message', (tester) async {
        await tester.pumpWidget(
          testApp(DsStateDisplay(type: DsStateType.error)),
        );
        final icon = tester.widget<Icon>(
          find.byKey(const ValueKey('ds-state-icon')),
        );
        expect(icon.icon, Icons.error_outline);
        final message = tester.widget<Text>(
          find.byKey(const ValueKey('ds-state-message')),
        );
        expect(message.data, 'Something went wrong');
      });

      testWidgets('renders custom message', (tester) async {
        await tester.pumpWidget(
          testApp(
            DsStateDisplay(
              type: DsStateType.error,
              message: 'Custom error',
            ),
          ),
        );
        final message = tester.widget<Text>(
          find.byKey(const ValueKey('ds-state-message')),
        );
        expect(message.data, 'Custom error');
      });

      testWidgets('renders action button', (tester) async {
        var actionFired = false;
        await tester.pumpWidget(
          testApp(
            DsStateDisplay(
              type: DsStateType.error,
              actionLabel: 'Try Again',
              onAction: () => actionFired = true,
            ),
          ),
        );
        final actionLabel = tester.widget<Text>(
          find.byKey(const ValueKey('ds-button-label')),
        );
        expect(actionLabel.data, 'Try Again');
        await tester.tap(find.byKey(const ValueKey('ds-state-action')));
        expect(actionFired, isTrue);
      });

      testWidgets('no action when actionLabel is null', (tester) async {
        await tester.pumpWidget(
          testApp(DsStateDisplay(type: DsStateType.error)),
        );
        expect(find.byKey(const ValueKey('ds-state-action')), findsNothing);
      });
    });

    group('custom icon', () {
      testWidgets('overrides default icon for empty state', (tester) async {
        await tester.pumpWidget(
          testApp(
            DsStateDisplay(
              type: DsStateType.empty,
              icon: Icons.cloud_off,
            ),
          ),
        );
        final icon = tester.widget<Icon>(
          find.byKey(const ValueKey('ds-state-icon')),
        );
        expect(icon.icon, Icons.cloud_off);
      });

      testWidgets('overrides default icon for error state', (tester) async {
        await tester.pumpWidget(
          testApp(
            DsStateDisplay(
              type: DsStateType.error,
              icon: Icons.warning,
            ),
          ),
        );
        final icon = tester.widget<Icon>(
          find.byKey(const ValueKey('ds-state-icon')),
        );
        expect(icon.icon, Icons.warning);
      });
    });
  });
}
